### PR TITLE
Draw images directly into the canvas when compositing for data purposes

### PR DESCRIPTION
**Pull request by eyebrowsoffire**
This fixes [110599](https://github.com/flutter/flutter/issues/110599)

Previously, the `BitmapCanvas` would always insert DOM elements for images. However, this doesn't work if the user is using the canvas to composite data (i.e. `_preserveImageData` is set). Make it so that we draw images directly into the canvas if this is set.

Also, if a data URL is requested from the canvas pool and we have no canvas, we have to create a canvas and create the data URL anyway, otherwise we end up with exceptions if the user hasn't actually drawn anything to the canvas.

### CODE REVIEW

**Review by yjbanov — APPROVED** [review]
lgtm

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 3.
View them at https://flutter-engine-gold.skia.org/cl/github/36058

File: lib/web_ui/lib/src/engine/html/bitmap_canvas.dart [hunk 1]
```diff
@@ -668,26 +678,40 @@ class BitmapCanvas extends EngineCanvas {
...
-      for (final DomElement clipElement in clipElements) {
-        rootElement.append(clipElement);
-        _children.add(clipElement);
-      }
+    if (_preserveImageData) {
+      // If we're preserving image data, we have to actually draw the image
+      // element onto the canvas.
+      // TODO(jacksongardner): Make this actually work with color filters.
+      setUpPaint(paint, null);
+      _canvasPool.drawImage(imgElement as DomHTMLImageElement, p);
```
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/html/bitmap_canvas.dart:686]: I don't know why all these methods just pass around `DomHTMLElement` when they all are `DomHTMLImageElement`. I guess I could clean that up.

[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/html/bitmap_canvas.dart:686]: Argh, there is actually one case where it's a div.

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 4.
View them at https://flutter-engine-gold.skia.org/cl/github/36058

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 6.
View them at https://flutter-engine-gold.skia.org/cl/github/36058

**Pull request merged**